### PR TITLE
Added '__slots__' in classes #29

### LINKED ISSUE #29: Add `__slots__` in classes
#### Description of the problem

Refer, http://book.pythontips.com/en/latest/__slots__magic.html

#### References/Other comments
This is issue is open for anyone to fix.

### PR DESCRIPTION
**Pull request by udhayacommits**
#### References to other Issues or PRs or Relevant literature
Fixes #29

#### Brief description of what is fixed or changed
Added __slots__ to all the classes
http://book.pythontips.com/en/latest/__slots__magic.html

### CODE REVIEW

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 1]
```diff
@@ -84,6 +84,8 @@ def peek(self):
 
 class ArrayStack(Stack):
 
+    __slots__ = ['maxsize', 'top', 'items']
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]: Add `dtype` to the list as well.

**Comment by czgdp1807** [comment]
In future PRs, make sure that you do not use `master` for making changes to the code. It can become quite complex to manage if your PR is closed or not accepted. For now this is fine and you can keep working on it. Thanks.

File: pydatastructs/miscellaneous_data_structures/stack.py (hunk 1 above)
Comment (czgdp1807) [reply pydatastructs/miscellaneous_data_structures/stack.py]: Don't make this change.
I will give you a `diff` by tonight, which you can apply on top of your changes and push them here.

Comment (udhayacommits) [reply pydatastructs/miscellaneous_data_structures/stack.py]: ok. I will wait for your changes.

**Comment by czgdp1807** [comment]
@udhayacommits Please apply the following diff on your master,
```diff
diff --git a/pydatastructs/linear_data_structures/arrays.py b/pydatastructs/linear_data_structures/arrays.py
index bf96ee5..cc92c93 100644
--- a/pydatastructs/linear_data_structures/arrays.py
+++ b/pydatastructs/linear_data_structures/arrays.py
@@ -63,9 +63,9 @@ class OneDimensionalArray(Array):
 
     .. [1] https://en.wikipedia.org/wiki/Array_data_structure#One-dimensional_arrays
     '''
-    
-    __slots__ = ['_size', '_data']
-    
+
+    __slots__ = ['_size', '_data', '_dtype']
+
     def __new__(cls, dtype=NoneType, *args, **kwargs):
         if dtype == NoneType or len(args) not in (1, 2):
             raise ValueError("1D array cannot be created due to incorrect"
diff --git a/pydatastructs/trees/binary_trees.py b/pydatastructs/trees/binary_trees.py
index a1d2aec..e8fb820 100644
--- a/pydatastructs/trees/binary_trees.py
+++ b/pydatastructs/trees/binary_trees.py
@@ -182,9 +182,7 @@ class BinarySearchTree(BinaryTree):
 
     pydatastructs.trees.binary_tree.BinaryTree
     """
-    
-    __slots__ = ['root_idx', 'comparator', 'tree', 'size']
-    
+
     def insert(self, key, data):
         walk = self.root_idx
         if self.tree[walk].key == None:
@@ -305,9 +303,6 @@ class AVLTree(BinarySearchTree):
     balance_factor = lambda self, node: self.right_height(node) - \
                                         self.left_height(node)
 
-    
-    __slots__ = ['root_idx', 'comparator', 'tree', 'size']
-    
     def _right_rotate(self, j, k):
         y = self.tree[k].right
         if y != None:
```
Use [pds_diff.txt](https://github.com/codezonediitj/pydatastructs/files/3763768/pds_diff.txt) if the above doesn't work.

**Comment by udhayacommits** [comment]
@czgdp1807 Please let me know how to apply the diff. I am only getting errors when trying using the text file.

**Comment by czgdp1807** [comment]
Are you using `git apply`? Can you please share the error message?

**Comment by udhayacommits** [comment]
@czgdp1807 , I was using TortoiseGit that did not work. But the command worked. thanks.

**Pull request merged**